src/client/java/teammates/client/scripts/sql/DataMigrationForCourseEntitySql.java

```diff
@@ -148,4 +186,200 @@ private Student createStudent(teammates.storage.sqlentity.Course newCourse,
 
         return newStudent;
     }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected void doOperation() {
+        log("Running " + getClass().getSimpleName() + "...");
+        log("Preview: " + isPreview());
+        setMigrationCriteria();
+
+        Cursor cursor = readPositionOfCursorFromFile().orElse(null);
+        if (cursor == null) {
+            log("Start from the beginning");
+        } else {
+            log("Start from cursor position: " + cursor.toUrlSafe());
+        }
+
+        boolean shouldContinue = true;
+        while (shouldContinue) {
+            shouldContinue = false;
+            Query<Course> filterQueryKeys = getFilterQuery().limit(BATCH_SIZE);
+            if (cursor != null) {
+                filterQueryKeys = filterQueryKeys.startAt(cursor);
+            }
+
+            QueryResults<?> iterator = filterQueryKeys.iterator();
+
+            Course currentCourse = null;
+            // Cascade delete the course if it is not fully migrated.
+            if (iterator.hasNext()) {
+                currentCourse = (Course) iterator.next();
+                if (currentCourse.isMigrated()) {
+                    currentCourse = (Course) iterator.next();
+                } else {
+                    deleteCourseCascade(currentCourse);
+                }
+            }
+
+            while (currentCourse != null) {
+                shouldContinue = true;
+                migrateWithoutTrx(currentCourse);
+                numberOfScannedKey.incrementAndGet();
+
+                cursor = iterator.getCursorAfter();
+                savePositionOfCursorToFile(cursor);
+
+                currentCourse = iterator.hasNext() ? (Course) iterator.next() : null;
+            }
+
+            if (shouldContinue) {
+                log(String.format("Cursor Position: %s", cursor.toUrlSafe()));
+                log(String.format("Number Of Course Entity Key Scanned: %d", numberOfScannedKey.get()));
+                log(String.format("Number Of Course Entity affected: %d", numberOfAffectedEntities.get()));
+                log(String.format("Number Of Course Entity updated: %d", numberOfUpdatedEntities.get()));
```

Comment:
Could we log this on a per-course basis instead? If the `BATCH_SIZE` / course is very big, we might not see this logging for hours. And we probably need to change the info being logged.

Reply:
Maybe we can keep this log for when a batch finishes, and perhaps log which course is being migrated before calling `doMigration`?

Reply:
Great idea, sounds good. Probably log once before and after each course?